design/ErrorHandling.md

```diff
@@ -80,90 +79,244 @@ device.addEventListener('validationerror', (event) => {
+    "invalid-state",  // `this` or an object argument was in a wrong state (e.g. mapped).
+    // TODO: more?
+    "out-of-memory"   // An allocation failed nonfatally.
+};
+
+interface GPUObjectStatus {
+    readonly attribute boolean valid;
+
+    // These are only set if valid == false.
+    readonly attribute GPUErrorType? type;
+    readonly attribute DOMString? message;
+};
+
+typedef (GPUBindGroup
+      or GPUBindGroupLayout
+      or GPUBuffer
+      or GPUCommandBuffer
+      or GPUCommandEncoder
+      or GPUComputePipeline
+      or GPUFence
+      or GPUPipelineLayout
+      or GPUProgrammablePassEncoder
+      or GPUQueue
+      or GPURenderPipeline
+      or GPUSampler
+      or GPUShaderModule
+      or GPUSwapChain
+      or GPUTexture
+      or GPUTextureView) GPUStatusableObject;
+
+partial interface GPUDevice {
+    Promise<GPUObjectStatus> getObjectStatus(GPUStatusableObject object);
 };
 ```
 
-(Note that object creation failures always send an error to the error log, regardless of the object type or the type of failure.)
+#### Alternatives
 
-If an application uses recoverable allocation, the implementation will still generate error log entries:
-a `"recoverable-out-of-memory"` error for the object creation, and `"validation-error"`s for any subsequent uses of the invalid object.
-The application may need to understand whether such error log entries were part of a recovered allocation (e.g. to avoid sending telemetry for those errors).
-To facilitate this filtering, the handle to the invalid object (including "expando" JavaScript properties) is attached to the error log entry (see above).
+- Instead of `device.getStatus(obj)`, have `obj.getStatus()`. Con: Pollutes objects with rarely-used methods. Pro: Maybe slightly more clear to users.
 
-### Recoverable errors in object creation
+- The Promise could reject if the object is invalid. Con: Promises can only reject to DOMExceptions, meaning we have to express error categories via established DOMException error codes. (Also not sure if DOMException allows custom `message`s.)
 
-A recoverable error is exposed as a `GPUObjectStatusQuery`.
+### Operation Status
 
-```
-// (Exact form/type subject to change.)
-typedef Promise<GPUObjectStatus> GPUObjectStatusQuery;
+For the "Void-Returning" operations, errors are reported via `getLastOperationStatus()`.
+
+  - `GPUQueue.submit`
+  - `GPUQueue.signal`
+  - `GPUBuffer.setSubData`
+  - `GPUBuffer.unmap`
+  - `GPUBuffer.destroy`
+  - `GPUTexture.destroy`
+  - `GPUFence.getCompletedValue` (never fails)
+
+`device.getLastOperationStatus()` always resolves to the status of the most recent void-returning operation.
+If called multiple times in a row, it returns the same result (but not the same object).
+
+`getLastOperationStatus` returns a Promise that resolves to a `GPUOperationStatus` object describing the status of the last operation on this object.
```

Comment:
For testing we would also like to make sure no errors happened in previous operations and not just the current operation.